
```{r, include = FALSE}
# enable python code previews; must use python 3
library(reticulate)
use_python("/usr/bin/python3")

ottrpal::set_knitr_image_path()
```

# BONUS AI for Bioinformatics

## Learning Objectives

-
-
-

## AI in Bioinformatics

[Bioinformatics](https://www.genome.gov/genetics-glossary/Bioinformatics) is an interdisciplinary field that combines biology, computer science, statistics, and data analysis to extract insights from biological data. It involves the development and application of computational tools, algorithms, and databases to store, organize, analyze, and interpret vast amounts of biological information, particularly in the context of genomics and public health.

Bioinformatics is especially important for research and work done with genomes. In genomics, bioinformatics enables the identification and annotation of genes, the prediction of protein structures and functions, and the analysis of genetic variations, such as single nucleotide polymorphisms and gene copy number variations. These analyses help in unraveling the genetic basis of diseases, discovering potential drug targets, and facilitating personalized medicine approaches.

The proliferation of genomic data in recent years have been transformative for the field of genomics. Advances in DNA sequencing technologies, particularly next-generation sequencing, have revolutionized the ability to generate vast amounts of genomic data quickly and at a reduced cost. This has led to an unprecedented increase in the availability and accessibility of genomic information. Researchers estimate that we will generate [between 2 and 40 exabytes of genomic data](https://www.genome.gov/about-genomics/fact-sheets/Genomic-Data-Science) within the next decade. Cloud computing platforms like [AnVIL](https://anvilproject.org/) are on the forefront of addressing the big genomic data storage and processing challenge.

You have seen principles in previous chapters that likewise apply to creating bioinformatics software. Creating bioinformatics software, however, comes with its own unique sets of challenges:

- **Data Complexity**: Genomic data is vast and complex, often consisting of large datasets with multiple dimensions, such as nucleotide sequences and quality scores, gene expression levels, and epigenetic modifications. Developing algorithms and software tools to efficiently handle and analyze this data requires expertise in data management, processing, and interpretation as well as the biological domain knowledge.

- **Diverse Data Types**: Bioinformatics software often handles diverse and complex data outside of strings and numeric types, such as [Bioconductor's `Single Cell Experiment`](https://www.bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html) data class. Knowledge of common data types used in bioinformatics is required to create functional software.

- **Algorithmic Complexity**: Developing accurate and efficient algorithms for various bioinformatics tasks, such as sequence alignment, gene prediction, and protein structure prediction, requires deep understanding of biological principles and computational techniques. Designing algorithms that can handle noise, errors, and biological variations is a persistent challenge.

- **Integration with Existing Workflows**: Bioinformatics software is often part of a larger analysis pipeline or workflow. Integrating AI-based tools seamlessly into existing bioinformatics pipelines and workflows, ensuring compatibility, and maintaining data integrity throughout the process requires special attention.

- **Ethical Considerations**: Generally, people are more sensitive about their genome or health data when compared to their Amazon purchases. Genomic data can be highly sensitive and AI tools are not typically designed with security and data sovereignty in mind.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Bioinformatics dino is excited to get started. It is wearing a lab coat and gloves and is saying 'Donâ€™t worry - We can use bioinformatics to fill in the blanks with frog DNA!'."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MCNeSO4aOm1iESWDLOGTcx3aLEbnu8UttV0QGVAeafE/edit#slide=id.g2313f95cd49_180_4")
```

## Ethics in Bioinformatics

When using AI to help generate code examples, perform refactoring, understand code, or create unit tests, you should keep the following in mind:

- **NEVER** submit chunks of code to AI tools without being 100% sure that they are free from [personally identifiable information (PII)](https://toolkit.ncats.nih.gov/glossary/personally-identifiable-information/). PII includes variablse like gender, race, birth date, geographic location, and disease diagnosis. If you need to, generate dummy data to test your code.

- **Be Reproducible**: Bioinformatics tools can be used for clinical decision making. It's important that anyone vetting or auditing your tool can trace any decision making steps.

## Hands-On Exercise

Now it's your turn to try. 

### The Code

You are working on improving an alignment algorithm. Alignments are typically used for comparing samples from individuals to a reference consensus sequence. To make sure things are working, you want to compare the alignment created by an existing function (`pairwiseAlignment`) to an alignment created by your tool (`my_new_Aligner`).

**Note**: This code is just an example and was written strictly for educational purposes.

```{r eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if (!requireNamespace("Biostrings", quietly = TRUE))
  BiocManager::install("Biostrings")

library(Biostrings)

reference <- DNAString("ATCGTACGTA")
query <- DNAString("ATCGTAGCTA")

# Perform pairwise sequence alignment
alignment <- pairwiseAlignment(reference, query)

# Perform sequence alignment using your algorithm
my_alignment <- my_new_Aligner(reference, query)

# Compare pairwise alignment to my_alignment
my_evaluation <- my_alignment_evaluator(reference, query, alignment, my_alignment)
```

### Questions

1. Things are going great with the toy data! You decide to make a tiny subset of your data for testing your aligner by copying the top 100 sequences from patient 1. You then submit the aligner and data to Bard to determine if there is anywhere handling this data can be made more efficient. Why is this problematic?

2. Where could you use AI tools to improve the code above?

## Summary

- 